persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DsePersistence.java

```diff
@@ -269,26 +269,40 @@ public Authenticator getAuthenticator() {
       boolean isTracingRequested,
       long queryStartNanoTime) {
     try {
-      org.apache.cassandra.service.QueryState internalState = Conversion.toInternal(state);
-      org.apache.cassandra.cql3.QueryOptions internalOptions = Conversion.toInternal(options);
-
       return Conversion.toFuture(
           Single.defer(
               () -> {
                 try {
-                  final UUID tracingId =
-                      beginTraceQuery(
-                          cql, internalState, internalOptions, customPayload, isTracingRequested);
+                  org.apache.cassandra.service.QueryState internalState =
+                      Conversion.toInternal(state);
+                  org.apache.cassandra.cql3.QueryOptions internalOptions =
+                      Conversion.toInternal(options);
 
                   checkIsLoggedIn(internalState);
 
                   CQLStatement statement = QueryProcessor.parseStatement(cql, internalState);
+
+                  // Please note that this needs to happen _before_ the beginTraceQuery, because
+                  // when we add bound values to the trace, we rely on the values having been
+                  // re-ordered by the following prepare (if named values were used that is).
                   internalOptions.prepare(statement.getBindVariables());
 
+                  final UUID tracingId =
+                      beginTraceQuery(
+                          cql, internalState, internalOptions, customPayload, isTracingRequested);
                   return processStatement(
-                      statement, state, options, customPayload, queryStartNanoTime, tracingId);
+                      statement,
+                      state,
+                      options,
+                      internalState,
+                      internalOptions,
+                      customPayload,
+                      queryStartNanoTime,
+                      tracingId);
                 } catch (Exception e) {
-                  return Single.error(Conversion.handleException(e));
+                  return Tracing.instance
+                      .stopSessionAsync()
+                      .andThen(Single.error(Conversion.handleException(e)));
```

Comment:
This `stopSessionAsync()` call on sync catch is confusing... looking at DSE codebase, it's added in `BatchMessage` and `QueryMessage` but not in `ExecuteMessage`.

Reply:
Great catch. Thanks!

Reply:
You're right. This doesn't make sense because it wouldn't be on the TPC thread processing the request so the thread local state for the tracing session would be invalid.